primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,794 @@
+
+/// Get the key for the value cache.
+///
+/// The key is build by hashing the actual storage `key` and the `storage_root`.
+fn value_cache_get_key(key: &[u8], storage_root: &impl AsRef<[u8]>) -> u64 {
+	let mut hasher = DefaultHasher::default();
+	hasher.write(key);
+	hasher.write(storage_root.as_ref());
+	hasher.finish()
+}
+
+/// The shared node cache.
+///
+/// Internally this stores all cached nodes in a [`LruCache`]. It ensures that when updating the
+/// cache, that the cache stays within it allowed bounds.
+struct SharedNodeCache<H: Hasher> {
+	/// The cached nodes, ordered by least recently used.
+	lru: LruCache<H::Out, NodeOwned<H::Out>>,
+	/// The size of [`Self::lru`] in bytes.
+	size_in_bytes: usize,
+	/// The maximum size in bytes of [`Self::lru`].
+	maximum_size_in_bytes: usize,
+}
+
+impl<H: Hasher> SharedNodeCache<H> {
+	/// Create a new instance.
+	fn new(maximum_size_in_bytes: usize) -> Self {
+		Self { lru: LruCache::unbounded(), size_in_bytes: 0, maximum_size_in_bytes }
+	}
+
+	/// Get the node for `key`.
+	///
+	/// This doesn't change the least recently order in the internal [`LruCache`].
+	fn get(&self, key: &H::Out) -> Option<&NodeOwned<H::Out>> {
+		self.lru.peek(key)
+	}
+
+	/// Update the cache with the `added` nodes and the `accessed` nodes.
+	///
+	/// The `added` nodes are the ones that have been collected by doing operations on the trie and
+	/// now should be stored in the shared cache. The `accessed` nodes are only referenced by hash
+	/// and represent the nodes that were retrieved from this shared cache through [`Self::get`].
+	/// These `accessed` nodes are being put to the front of the internal [`LruCache`] like the
+	/// `added` ones.
+	///
+	/// After the internal [`LruCache`] was updated, it is ensured that the internal [`LruCache`] is
+	/// inside its bounds ([`Self::maximum_size_in_bytes`]).
+	fn update(
+		&mut self,
+		added: impl IntoIterator<Item = (H::Out, NodeOwned<H::Out>)>,
+		accessed: impl IntoIterator<Item = H::Out>,
+	) {
+		accessed.into_iter().for_each(|key| {
+			// Access every node in the lru to put it to the front.
+			self.lru.get(&key);
+		});
+		added.into_iter().for_each(|(key, node)| {
+			self.size_in_bytes += key.as_ref().len() + node.size_in_bytes();
+
+			if let Some((r_key, r_node)) = self.lru.push(key, node) {
+				self.size_in_bytes = self
+					.size_in_bytes
+					.saturating_sub(r_key.as_ref().len() + r_node.size_in_bytes());
+			}
+		});
+
+		while self.size_in_bytes > self.maximum_size_in_bytes {
+			// This should always be `Some(_)`, otherwise something is wrong!
+			if let Some((key, node)) = self.lru.pop_lru() {
+				self.size_in_bytes =
+					self.size_in_bytes.saturating_sub(key.as_ref().len() + node.size_in_bytes());
+			}
+		}
+	}
+}
+
+/// Configuration of the [`SharedTrieCache`].
+#[derive(Debug, Clone)]
+pub struct Configuration {
+	/// The maximum size in bytes the cache should use.
+	pub maximum_size_in_bytes: usize,
```

Comment:
Hmm... I think this should probably be renamed, since it's not actually a pure maximum, so it's a little confusing. This essentially controls two things: the maximum amount of memory that the shared LRU cache is roughly going to use, and the amount of memory which is going to be preallocated for the value cache. So you can't actually set it to e.g. `usize::MAX` to get an unbound cache, as that will result in a crash due to an allocation failure. Maybe rename it to `cache_size_in_bytes` or something like that?